modules/fhir-structure/src/blaze/fhir/spec.clj

```diff
@@ -46,16 +47,99 @@
+
+
+(defn fhir-path
+  "Given a vector `path` and a `resource`, returns the fhir-path as String for
+  the given path."
+  [path resource]
+  (subs (reduce (fn [x y] (str x (if (keyword? y)
+                                   (format ".%s" (name y))
+                                   (format "[%s]" y)))) ""
+                (fhir-path-data path resource)) 1))
+
+
+(defn get-regex
+  "Returns the first regex pattern in `form`."
+  [form]
+  (clojure.walk/postwalk
+    #(if (sequential? %)
+       (some identity %)
+       (when (= (type %) Pattern) %))
+    form))
+
+
+(defn get-diagnostics-from-problem
+  "Returns a diagnostics message from the given arguments pred, val and via.
+  pred: A Function to evaluate the given val.
+  val: Error Value.
+  via: Vector-path of fhir types to the given error.
+  If pred contains Regex-patetrns to check, the first is included in diagnostics
+  message."
+  [pred val via]
+  (if-let [regex (get-regex pred)]
+    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
+            val (name (last via)) regex)
+    (format "Error on Value `%s`. Expected type is `%s`."
+            val (name (last via)))))
+
+
 (defn explain-data
-  {:arglists '([resource])}
-  [{type :resourceType :as resource}]
-  (when-let [spec (s2/get-spec (keyword "fhir" type))]
-    (s2/explain-data spec resource)))
+  "Given a resource witch includes :resourceType key and severity as string,
+  returns nil if the resource is conform to the spec :fhir/resourceType,
+  else a map with at least the keys :fhir/issues whoes value is a collection of
+  issue-maps to build a OperationsOutcome and ::problems whose value is a
+  collection of problem-maps, where problem-map has at least :path :pred and
+  :val keys describing the predicate and the value that failed at that path."
+  {:arglists '([resource] [resource severity])}
+  ([{type :resourceType :as resource}] (explain-data resource "error"))
+  ([{type :resourceType :as resource} severity]
+   (if type
+     (when-let
+       [spec-error (when-let
+                     [spec (s2/get-spec (keyword "fhir" type))]
+                     (s2/explain-data spec resource))]
+       (assoc
+         spec-error
+         :fhir/issues
+         (into
+           []
+           (map (fn [problem]
+                  (let [{:keys [val via in pred]} problem]
+                    {:fhir.issues/severity
+                     severity
+                     :fhir.issues/code
+                     "invariant"
+                     :fhir.issues/diagnostics
+                     (get-diagnostics-from-problem pred val via)
+                     :fhir.issues/expression
+                     (fhir-path in resource)}))
+                (:clojure.spec.alpha/problems spec-error)))))
+     {:fhir/issues
+      (conj
+        []
+        {:fhir.issues/severity "error"
+         :fhir.issues/code "value"
+         :fhir.issues/diagnostics
+         "Given resource does not contain :resourceType key!"})})))
```

Comment:
Please output the same message, that would be output if another property would be missing. We like to never expose internal things like keywords.